packages/api/portfolioDrafts/application/createApplicationStep.ts

```diff
@@ -0,0 +1,146 @@
+import { APPLICATION_STEP } from "../../models/PortfolioDraft";
+import { dynamodbDocumentClient as client } from "../../utils/dynamodb";
+import {
+  DATABASE_ERROR,
+  NO_SUCH_PORTFOLIO_DRAFT,
+  PATH_PARAMETER_REQUIRED_BUT_MISSING,
+  REQUEST_BODY_EMPTY,
+  REQUEST_BODY_INVALID,
+} from "../../utils/errors";
+import { ApiSuccessResponse, SuccessStatusCode, ValidationErrorResponse } from "../../utils/response";
+import { isApplicationStep, isValidJson, isValidUuidV4 } from "../../utils/validation";
+export interface ApplicationValidationError {
+  applicationName: string;
+  invalidParameterName: string;
+  invalidParameterValue: string;
+  validationMessage: ValidationMessage;
+}
+
+/**
+ * Submits the Application Step of the Portfolio Draft Wizard
+ *
+ * @param event - The POST request from API Gateway
+ */
+export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
+  const portfolioDraftId = event.pathParameters?.portfolioDraftId;
+  if (!portfolioDraftId) {
+    return PATH_PARAMETER_REQUIRED_BUT_MISSING;
+  }
+  if (!isValidUuidV4(portfolioDraftId)) {
+    return NO_SUCH_PORTFOLIO_DRAFT;
+  }
+  if (!event.body) {
+    return REQUEST_BODY_EMPTY;
+  }
+  if (!isValidJson(event.body)) {
+    return REQUEST_BODY_INVALID;
+  }
+  const requestBody = JSON.parse(event.body);
+  if (!isApplicationStep(requestBody)) {
+    return REQUEST_BODY_INVALID;
+  }
+  const applicationStep: ApplicationStep = requestBody;
+  const errors: Array<ApplicationValidationError> = validateApplicationStep(applicationStep);
+  if (errors.length) {
+    return createValidationErrorResponse({ input_validation_errors: errors });
+  }
+  try {
+    await client.send(
+      new UpdateCommand({
+        TableName: process.env.ATAT_TABLE_NAME ?? "",
+        Key: {
+          id: portfolioDraftId,
+        },
+        UpdateExpression: "set #portfolioVariable = :application, updated_at = :now",
+        ExpressionAttributeNames: {
+          "#portfolioVariable": APPLICATION_STEP,
+        },
+        ExpressionAttributeValues: {
+          ":application": applicationStep,
+          ":now": new Date().toISOString(),
+        },
+        ConditionExpression: "attribute_exists(created_at)",
+        ReturnValues: "ALL_NEW",
+      })
+    );
+  } catch (error) {
+    if (error.name === "ConditionalCheckFailedException") {
+      return NO_SUCH_PORTFOLIO_DRAFT;
+    }
+    console.error("Database error: " + error);
+    return DATABASE_ERROR;
+  }
+  return new ApiSuccessResponse<ApplicationStep>(applicationStep, SuccessStatusCode.CREATED);
+}
+
+/**
+ * Validates the given ApplicationStep
+ * Performs input validation on all Applications and Environments contained therein
+ * @param applicationStep the object to validate
+ * @returns a collection of application validation errors
+ */
+export function validateApplicationStep(applicationStep: ApplicationStep): Array<ApplicationValidationError> {
```

Comment:
I think these functions (`validateApplicationStep`, `validateApplication`, `validateEnvironment`) should be living in `validation.ts` . `createValidationErrorResponse` might be fine to keep in this fn, if its renamed.

Reply:
I believe there is some confusion due to overloaded use of the words _valid_ and _validate_ amongst ourselves.

I think as a team we need to adopt shared terminology for:
1. testing and evaluating the shape of an object and whether it meets an interface
2. performing tests on specific data within an object

I'll start a slack thread for this.

As to the specifics above, all of these are for the latter and involving input validation for data inside the objects.
I see `validation.ts` as performing the former and I don't believe these functions belong there.  Perhaps I need to avoid using the word _validate_ for these functions.  Or maybe I need to expand my interpretation of what belongs in `validation.ts`

In either case I think we should discuss these issues as a team.

Reply:
I agree

Reply:
Preliminary result of that discussion has been to be purposeful in the distinction and organization of the _object validation_ and `data validation` solutions.

Reply:
Renamed functions in 3db4345f993390b575b222786ecdcf8303b2d149